# Chapter 3 - Intro to neural prediction | Forward propagation 

## First NN, 
- 1 input 
- 1 output
- 1 weight

In [2]:
weight = 0.1
def neural_network(input, weight):
    prediction = input * weight
    return prediction

In [3]:
number_of_toes = [8.5, 9.5, 10, 9]
input= number_of_toes[0]
prediction = neural_network(input, weight)
print(prediction)

0.8500000000000001


## Making a prediction with multiple inputs

In [4]:
# Preforming a weighted sum of the inputs
def w_sum(a, b):
    # a is the input, b is the weight
    assert(len(a) == len(b))
    output = 0
    for i in range(len(a)):
        output += (a[i] * b[i])
    
    return output

weights = [0.1, 0.2, 0]
def neural_network(input, weights):
    pred = w_sum(input, weights)
    return pred

In [ ]:
toes = [8.5, 9.5, 9.9, 9.0]
wlrec = [0.65, 0.8, 0.8, 0.9]
nfans = [1.2, 1.3, 0.5, 1.0]

input = [toes[0], wlrec[0], nfans[0]]
pred = neural_network(input, weights)

In [16]:
def elementwise_multiplication(number, vector):
    res_vec = [0, 0, 0]

    assert(len(res_vec) == len(vector))

    for i in range(len(vector)):
        res_vec[i] = number *vector[i]
    
    return res_vec

def elementwise_addition(vec_a, vec_b):
    assert(len(vec_a) == len(vec_b))
    res_vec = []
    
    for i in range(len(vec_a)):
        res_vec.append(vec_a[i] + vec_b[i])

    return res_vec

In [8]:
vec_a = [1, 2, 3]
vec_b = [4, 5, 6]

print(elementwise_multiplication(vec_a, vec_b))
print(elementwise_addition(vec_a, vec_b))

[4, 10, 18]
[5, 7, 9]


## Using numpy

In [1]:
import numpy as np

In [9]:

weights = np.array([0.1,0.2,0])

def neural_network(input:np.ndarray , weights: np.ndarray):
    pred = input.dot(weights)

    return pred


In [11]:
toes = np.array([8.5, 9.5, 9.9, 9.0])
wlrec = np.array([0.65, 0.8, 0.8, 0.9])
nfans = np.array([1.2, 1.3, 0.5, 1.0])

input = np.array([toes[0], wlrec[0], nfans[0]])
pred = neural_network(input, weights)
print(pred)

0.9800000000000001


## Making a prediction with multiple outputs

In [12]:
weights = [0.3, 0.2, 0.9]

def neural_network(input, weights):
    pred = elementwise_multiplication(input, weights) # Multiply the input with each weight
    return pred

In [17]:
wlrec = [0.65, 0.8, 0.8, 0.9]
input = wlrec[0]

pred = neural_network(input, weights)
print(pred)

[0.195, 0.13, 0.5850000000000001]


- 0.195 hurt prediction
- 0.13 win prediction
- 0.58 sad prediction

## Prediction with multiple inputs and outputs

In [6]:
           # In1, In2, In3 
weights = [[0.1, 0.1, -0.3], # first weights for all features 
           [0.1 , 0.2, 0.0], # 2nd for all features
           [0.0, 1.3, 0.1]] # 3d for all features

def w_sum(input, weights):
    assert(len(input) == len(weights))
    output = 0

    for i in range(len(input)):
        output+= (input[i]*weights[i]) 
    return output

def vec_mat_mul(vector, matrix):
    """
    Performs matrix multiplication between a vector and a matrix
    Take the input and do a w_sum for each weigh dimension
    It performs three independent weighted sums of the input to make three predictions.
    """
     
    assert(len(vector) == len(matrix))

    output = [0,0,0]

    for i in range(len(vector)):
        output[i] = w_sum(vector, matrix[i])
    
    return output
    

def neural_network(input, weights):

    pred = vec_mat_mul(input, weights)

    return pred

In [2]:
toes = [8.5, 9.5, 9.9, 9.0]
wlrec = [0.65, 0.8, 0.8, 0.9]
nfans = [1.2, 1.3, 0.5, 1.0]

In [7]:
input = [toes[0], wlrec[0], nfans[0]]

pred = neural_network(input, weights)
print(pred)

[0.555, 0.9800000000000001, 0.9650000000000001]


## Predicting on predictions

In [8]:
ih_wgt = [[0.1, 0.2, -0.1], 
           [-0.1 , 0.1, 0.9], 
           [0.1, 0.4, 0.1]]
hp_wgt = [[0.3, 1.1, -0.3], 
           [0.1 , 0.2, 0.0], 
           [0.1, 1.3, 0.1]]

weights =  [ih_wgt, hp_wgt]

In [9]:
def neural_network(input, weights):

    hid = vec_mat_mul(input, weights[0])
    pred = vec_mat_mul(hid, weights[1])
    return pred

In [10]:
input = [toes[0], wlrec[0], nfans[0]]
pred = neural_network(input, weights)

print(pred)

[0.21350000000000002, 0.14500000000000002, 0.5924999999999999]


### Numpy version

In [13]:
ih_wgt = np.array([
            [0.1, 0.2, -0.1], 
            [-0.1 , 0.1, 0.9], 
            [0.1, 0.4, 0.1]])

hp_wgt = np.array(
            [[0.3, 1.1, -0.3], 
            [0.1 , 0.2, 0.0], 
            [0.1, 1.3, 0.1]])

weights = [ih_wgt, hp_wgt]

In [14]:
def neural_network(input, weights):

    hid = input.dot(weights[0])
    pred = hid.dot(weights[1])
    return pred

In [16]:
toes = np.array([8.5, 9.5, 9.9, 9.0])
wlrec = np.array([0.65, 0.8, 0.8, 0.9])
nfans = np.array([1.2, 1.3, 0.5, 1.0])

input = np.array([toes[0], wlrec[0], nfans[0]])

pred = neural_network(input, weights)
print(pred)

[ 0.4815  1.256  -0.286 ]
